src/nodes.coffee

```diff
@@ -1069,6 +1084,106 @@ exports.Arr = class Arr extends Base
 
 #### Class
 
+# A node representing a Javascript class.
+# Initialize a **JsClass** with its name, an optional superclass, and a
+# list of methods.
+exports.JsClass = class JsClass extends Base
+  constructor: (@name, @parent, @methods = []) ->
+    @boundMethods = []
+
+  isStatement: YES
+
+  setConstructor: (source, constructor) ->
+    source.error 'cannot define more than one constructor in a class' if @ctor
+
+    if constructor instanceof Code
+      @ctor = constructor
+    else
+      @generatedCtor = true
+
+      apply = new Value constructor, [new Access new PropertyName 'apply']
+      @ctor = new Code
+      @ctor.body.push new Literal 'super(...arguments)' if @parent
+      @ctor.body.push new Call apply, [new ThisLiteral, new IdentifierLiteral 'arguments']
+      @ctor.body.makeReturn()
+
+    @ctor.ctor = true
+    @ctor.noReturn = true
+    @ctor.isDerivedCtor = @parent?
+
+    name = new PropertyName 'constructor'
+    @addMethod name, @ctor
+
+  addMethod: (name, method) ->
+    method.name = name
+    method.isMethod = true
+
+    if method.bound and not method.static
+      method.bound = false
+      @boundMethods.push name
+
+    # Override `method`'s `compileToFragments` to append to the class's method fragments
+    # This allows `method.compileToFragments` to be called with the correct scope whilst still
+    # being generated in the class body.
+    compileToFragments = method.compileToFragments
+    method.compileToFragments = (o) =>
+      o.indent += TAB
+      @methodFragments.push compileToFragments.apply(method, arguments)
+      o.indent = o.indent[0..-1 - TAB.length]
+      []
+
+    @methods.push method
+    method
+
+  addStaticMethod: (name, method) ->
+    method.static = true
+    @addMethod name, method
+
+  compileNode: (o) ->
+    { indent } = o
+    o.indent += TAB
+
+    @bindMethods o
+
+    code = []
+    code.push @makeCode "#{indent}class "
+    code.push @name.compileNode(o)...
+    code.push @makeCode ' '
+    if @parent
+      code.push @makeCode 'extends '
+      code.push @parent.compileNode(o)...
+      code.push @makeCode ' '
+    code.push @makeCode '{'
+    if @methods.length
+      @bodyFragment    = @makeCode ''
+      @methodFragments = []
+      code.push @makeCode("\n")
+      code.push @ctor.compileToFragments(o)... if @generatedCtor
+      code.push @bodyFragment, @makeCode("\n#{indent}")
+    code.push @makeCode '}'
+
+    code
+
+  bindMethods: (o) ->
+    # If there are no bound methods, there's nothing to do
+    return if @boundMethods.length is 0
+
+    # If there's no ctor defined we need to generate one
+    if not @ctor
+      @generatedCtor = true
+      ctor = new Code
+      @setConstructor null, ctor
+
+    for name in @boundMethods
+      lhs = new Value(new ThisLiteral, [name]).compile o
+      @ctor.body.unshift new Literal "#{lhs} = #{utility 'bind', o}(#{lhs}, this)"
+
+    ctor?.body.unshift new Literal 'super(...arguments)' if @parent
+
+  finalize: ->
```

Comment:
The second part of the method hoisting hack. Once a class body has been compiled, `finalize()` is called to inject the compiled methods into the `JsClass` body.